clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/names/EcsServerGroupNameResolver.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2020 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.clouddriver.ecs.names;
+
+import com.amazonaws.services.ecs.AmazonECS;
+import com.amazonaws.services.ecs.model.*;
+import com.google.common.collect.Lists;
+import com.netflix.spinnaker.moniker.Moniker;
+import com.netflix.spinnaker.moniker.Namer;
+import java.util.*;
+import java.util.stream.IntStream;
+
+public class EcsServerGroupNameResolver {
+
+  private static final int SEQUENTIAL_NUMBERING_NAMESPACE_SIZE = 1000;
+
+  private static final int MAX_NEXT_SERVER_GROUP_ATTEMPTS = 5;
+
+  private final String ecsClusterName;
+  private final AmazonECS ecs;
+  private final String region;
+  private final Namer<EcsResource> naming;
+
+  public EcsServerGroupNameResolver(
+      String ecsClusterName, AmazonECS ecs, String region, Namer<EcsResource> namer) {
+    this.ecsClusterName = ecsClusterName;
+    this.ecs = ecs;
+    this.region = region;
+    this.naming = namer;
+  }
+
+  public EcsServerGroupName resolveNextName(String application, String stack, String detail) {
+    Moniker m = Moniker.builder().app(application).detail(detail).stack(stack).build();
+
+    return resolveNextName(m);
+  }
+
+  public EcsServerGroupName resolveNextName(Moniker currentName) {
+    Set<Integer> takenSequences = new HashSet<>();
+
+    // 1. Get a list of all of the services
+    List<String> allServices = listAllServices(ecsClusterName);
+
+    // 2. Get the details of the services in the same server group
+    List<List<String>> serviceBatches = Lists.partition(allServices, 10);
+    for (List<String> serviceBatch : serviceBatches) {
+      DescribeServicesRequest request =
+          new DescribeServicesRequest()
+              .withCluster(ecsClusterName)
+              .withServices(serviceBatch)
+              .withInclude("TAGS");
+      DescribeServicesResult result = ecs.describeServices(request);
+      for (Service service : result.getServices()) {
+        Moniker moniker = naming.deriveMoniker(new EcsResourceService(service));
+
+        if (Objects.equals(currentName.getApp(), moniker.getApp())
```

Comment:
If this is a string, might make more sense to use StringUtils.equals instead of Objects.equals

Reply:
The contract between `Objects.equals(String, String)` and `StringUtils.equals(String, String)` is the same as they both end up just calling `String.equals()` but it's probably clearer to use the `StringUtils` one here.

Reply:
Agreed, but it is also possible that the hashcode is different, causing the equals operator on two string objects to return false (even if both strings are identical)